### PR TITLE
NETOBSERV-1473: migrating netobserv-agent to TCx for new kernels

### LINKED ISSUE #230: migrate the existing TC hooks to TCX
Now cilium supports TCX hooks 
https://github.com/cilium/ebpf/commit/417f8a264d2e691649ede9b995fb334e6c58ca71
The ebpf agent need to follow https://github.com/cilium/cilium/issues/27632 to migrate to tcx and remove dependency on `netlink` and `Qdisc`

reference: http://vger.kernel.org/bpfconf2023_material/tcx_meta_netdev_borkmann.pdf

### PR DESCRIPTION
**Pull request by msherif1234**
## Description
Migrate ebpf agent from TC to TCX "TC eXpress" if the kernel support it, this PR fix #230.

- how to check TCX is enabled in ur kernel
```sh
zgrep CONFIG_NET_XGRESS /boot/config-$(uname -r)
CONFIG_NET_XGRESS=y
```
- how to check if TCX hooks are attached to all interfaces
```sh
sudo bpftool net

xdp:

tc:
wlp0s20f3(2) tcx/ingress tcx_ingress_flow_parse prog_id 3923 link_id 214 
wlp0s20f3(2) tcx/egress tcx_egress_flow_parse prog_id 3922 link_id 213 
br-68b237cee3d3(3) tcx/ingress tcx_ingress_flow_parse prog_id 3923 link_id 216 
br-68b237cee3d3(3) tcx/egress tcx_egress_flow_parse prog_id 3922 link_id 215 
docker0(4) tcx/ingress tcx_ingress_flow_parse prog_id 3923 link_id 218 
docker0(4) tcx/egress tcx_egress_flow_parse prog_id 3922 link_id 217 
enp9s0u2u1u2(1736) tcx/ingress tcx_ingress_flow_parse prog_id 3923 link_id 220 
enp9s0u2u1u2(1736) tcx/egress tcx_egress_flow_parse prog_id 3922 link_id 219 

flow_dissector:

netfilter:

```
- testing 
* [X] using `6.6.2-201.fc39.x86_64` run standalone agent code as well as e2e test and both passed with the new TCX. Waiting for OCP build with rhel9.4 to be able to resume the testing on OCP cluster

* [ ] pkt drop
* [ ] DNS
* [ ] RTT
* [ ] SRIOV secondary interface
* [ ] PCA
* [ ] performance and scale test

* [X] on older kernel we will get this warning and fall back to the legacy TC hooks
```sh
time="2024-02-05T12:56:55Z" level=info msg="interface detected. trying to attach TCX hook" component=agent.Flows interface="{1b684d9681b6cb8 106 NS(none)}"
time="2024-02-05T12:56:55Z" level=warning msg="can't attach to TCx hook flow ebpfFetcher. fall back to use legacy TC hook" component=agent.Flows error="failed to attach TCX egress: tcx not supported (requires >= v6.6)" interface="{1b684d9681b6cb8 106 NS(none)}"
```
Note: GH use `ubuntu-latest` which is currently `6.2.0` so doesn't have TCX support yet
## Dependencies

OCP with rhel9.4 kernel

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [ ] Will this change affect NetObserv / Network Observability operator? If not, you can ignore the rest of this checklist.
* [X] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [X] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [X] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

**Comment by msherif1234** [comment]
verified with https://brewweb.engineering.redhat.com/brew/taskinfo?taskID=59954477

File: pkg/ebpf/tracer.go [hunk 1]
```diff
@@ -377,6 +421,19 @@ func (m *FlowFetcher) Close() error {
 	if len(errs) == 0 {
 		return nil
 	}
+	for iface, l := range m.egressTCXLink {
+		log := log.WithField("interface", iface)
+		log.Debug("detach egress TCX hook")
+		l.Close()
+
```
Comment (jpinsonneau) [inline pkg/ebpf/tracer.go:428]:
```suggestion
```

**Review by jpinsonneau — APPROVED** [review]
Code looks good; any tips to test this on a cluster ?

**Comment by msherif1234** [comment]
> Code looks good; any tips to test this on a cluster ?

yes detailed instructions on how to test this on OCP is in the comment section for 
https://issues.redhat.com/browse/NETOBSERV-1473

**Comment by msherif1234** [comment]
did multiple rebase to keep with latest code

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -509,10 +510,14 @@ func (f *Flows) onInterfaceAdded(iface ifaces.Interface) {
...
 		return
 	}
-	alog.WithField("interface", iface).Info("interface detected. Registering flow ebpfFetcher")
-	if err := f.ebpf.Register(iface); err != nil {
+	alog.WithField("interface", iface).Info("interface detected. trying to attach TCX hook")
+	if err := f.ebpf.AttachTCX(iface); err != nil {
 		alog.WithField("interface", iface).WithError(err).
-			Warn("can't register flow ebpfFetcher. Ignoring")
-		return
+			Warn("can't attach to TCx hook flow ebpfFetcher. fall back to use legacy TC hook")
```
Comment (jotak) [inline pkg/agent/agent.go:516]: nit: I would use an Info log rather then Warn as TCx still won't be supported in many configs hence that's still a "normal" behaviour in most cases? (so it doesn't scare the users :-) )

File: pkg/ebpf/tracer.go [hunk 3]
```diff
@@ -196,10 +199,44 @@ func NewFlowFetcher(cfg *FlowFetcherConfig) (*FlowFetcher, error) {
...
 }
 
+func (m *FlowFetcher) AttachTCX(iface ifaces.Interface) error {
+	ilog := log.WithField("iface", iface)
+	if iface.NetNS != netns.None() {
+		if err := unix.Setns(int(iface.NetNS), unix.CLONE_NEWNET); err != nil {
+			return fmt.Errorf("failed to setns to %s: %w", iface.NetNS, err)
+		}
+	}
+	egrLink, err := link.AttachTCX(link.TCXOptions{
```
Comment (jotak) [inline pkg/ebpf/tracer.go:215]: I guess this function should honor the `FlowFetcher` `enableIngress` and `enableEgress` flags, like `Register` does?

Also that makes me think, as the agent can already be configured to capture only ingress or only egress via `config.Direction`, the filter `config.FlowFilterDirection` recently added was probably not necessary? Can we remove it? (not asking that for this PR)

Comment (jotak) [reply pkg/ebpf/tracer.go]: > Also that makes me think, as the agent can already be configured to capture only ingress or only egress via config.Direction, the filter config.FlowFilterDirection recently added was probably not necessary? Can we remove it? (not asking that for this PR)

Ok forget about it .. as discussed in slack, these two settings are redundant in ACCEPT filter mode, but they are not in REJECT mode. E.g. we may capture all ingress+egress globally but filter out a small subset of flows that are ingress or egress, for example on a particular port or subnet.

**Review by jotak — APPROVED** [review]
code lgtm!

**Comment by msherif1234** [comment]
discussed with @Amoghrd and he is fine running QE verification post merge

**Pull request merged**